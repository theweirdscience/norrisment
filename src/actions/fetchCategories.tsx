import { Action, ActionFactory } from ".";

export const FETCH_CATEGORIES_REQUEST = "FETCH_CATEGORIES_REQUEST";

export const fetchCategoriesRequest: ActionFactory = (): Action => {
  return {
    type: FETCH_CATEGORIES_REQUEST
  };
};

export const FETCH_CATEGORIES_COMPLETE = "FETCH_CATEGORIES_COMPLETE";

export const fetchCategoriesComplete: ActionFactory = (data: any): Action => {
  return {
    type: FETCH_CATEGORIES_COMPLETE,
    payload: data
  };
};
