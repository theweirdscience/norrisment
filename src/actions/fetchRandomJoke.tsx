import { Action, ActionFactory } from ".";

export const FETCH_RANDOM_JOKE_REQUEST = "FETCH_RANDOM_JOKE_REQUEST";

export const fetchRandomJokeRequest: ActionFactory = (data: any): Action => {
  return {
    type: FETCH_RANDOM_JOKE_REQUEST,
    payload: data
  };
};

export const FETCH_RANDOM_JOKE_COMPLETE = "FETCH_RANDOM_JOKE_COMPLETE";

export const fetchRandomJokeComplete: ActionFactory = (data: any): Action => {
  return {
    type: FETCH_RANDOM_JOKE_COMPLETE,
    payload: data
  };
};
