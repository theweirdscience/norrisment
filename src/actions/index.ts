export interface Action {
  type: string;
  payload?: any;
}

export interface ActionFactory {
  (any?): Action;
}

export * from "./fetchCategories";
export * from "./fetchRandomJoke";
