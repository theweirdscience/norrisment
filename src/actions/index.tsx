export interface Action {
  type: string;
  payload?: any;
}

export interface ActionFactory {
  (data?: any): Action;
}

export * from "./fetchCategories";
export * from "./fetchRandomJoke";
