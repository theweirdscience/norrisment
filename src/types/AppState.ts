export interface Joke {
  categories: string[];
  icon_url: string;
  id: string;
  url: string;
  value: string;
}

export interface Category {
  name: string;
}

export interface Settings {
  requestInProgress: boolean;
}

export interface AppState {
  settings: Settings;
  categories: Category[];
  randomJoke?: Joke;
}
