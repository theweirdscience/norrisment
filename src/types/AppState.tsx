
export interface Norrisism {
  id: string
  question: string
  answer: string
}

export interface AppState {
  norrisisms: Norrisism[]
}
