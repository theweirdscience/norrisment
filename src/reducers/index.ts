import { Category, AppState } from "../types";
import {
  Action,
  FETCH_CATEGORIES_REQUEST,
  FETCH_CATEGORIES_COMPLETE,
  FETCH_RANDOM_JOKE_COMPLETE,
  REQUEST_IN_PROGRESS_START,
  REQUEST_IN_PROGRESS_FINISH,
  CLEAR_RANDOM_JOKE
} from "../actions";
import * as FontAwesome from "@fortawesome/fontawesome";
import {
  faCode,
  faFilm,
  faStroopwafel,
  faUserCircle,
  faFlask,
  faFutbol,
  faBalanceScale,
  faChurch,
  faCrow,
  faHistory,
  faMusic,
  faPlane,
  faBriefcase,
  faMoneyBillAlt,
  faTshirt
} from "@fortawesome/fontawesome-free-solid";

FontAwesome.library.add(
  faCode,
  faFilm,
  faStroopwafel,
  faUserCircle,
  faFlask,
  faFutbol,
  faBalanceScale,
  faChurch,
  faCrow,
  faHistory,
  faMusic,
  faPlane,
  faBriefcase,
  faMoneyBillAlt,
  faTshirt
);

const defaultState: AppState = {
  settings: {
    requestInProgress: false
  },
  categories: []
};

const iconForCategoryMap: { [key: string]: string } = {
  dev: "code",
  movie: "film",
  food: "stroopwafel",
  celebrity: "user-circle",
  science: "flask",
  sport: "futbol",
  political: "balance-scale",
  religion: "church",
  animal: "crow",
  history: "history",
  music: "music",
  travel: "plane",
  career: "briefcase",
  money: "money-bill-alt",
  fashion: "tshirt"
};
function getIconForCategory(name: string): string {
  return iconForCategoryMap[name] ? iconForCategoryMap[name] : "";
}

const requestInProgressReducer = (state: AppState = defaultState, action: Action) => {
  switch (action.type) {
    case REQUEST_IN_PROGRESS_START:
      return {
        ...state,
        settings: {
          ...state.settings,
          requestInProgress: true
        }
      };

    case REQUEST_IN_PROGRESS_FINISH:
      return {
        ...state,
        settings: {
          ...state.settings,
          requestInProgress: false
        }
      };

    default:
      return state;
  }
};

const categoriesReducer = (state: AppState = defaultState, action: Action) => {
  switch (action.type) {
    case FETCH_CATEGORIES_COMPLETE:
      return {
        ...state,
        categories: action.payload
          .filter(name => name !== "explicit")
          .map(name => ({ name: name }))
          .map(category => ({ ...category, icon: getIconForCategory(category.name) }))
      };

    default:
      return state;
  }
};

const randomJokeReducer = (state: AppState = defaultState, action: Action) => {
  switch (action.type) {
    case FETCH_RANDOM_JOKE_COMPLETE:
      return {
        ...state,
        randomJoke: action.payload
      };

    case CLEAR_RANDOM_JOKE:
      return {
        ...state,
        randomJoke: null
      };

    default:
      return state;
  }
};

const mergeReducers = reducers => (state: AppState, action: Action) => {
  reducers.forEach(reducer => {
    state = reducer(state, action);
  });
  return state;
};

export default mergeReducers([categoriesReducer, randomJokeReducer, requestInProgressReducer]);
