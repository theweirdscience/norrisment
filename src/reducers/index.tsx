import { Category, AppState } from "../types";
import { Action, FETCH_CATEGORIES_REQUEST, FETCH_CATEGORIES_COMPLETE, FETCH_RANDOM_JOKE_COMPLETE } from "../actions";

const defaultState: AppState = {
  categories: [],
  jokes: []
};

const categoriesReducer = (state: AppState = defaultState, action: Action) => {
  switch (action.type) {
    case FETCH_CATEGORIES_COMPLETE:
      const result = {
        ...state,
        categories: action.payload.map(name => ({ name: name }))
      };
      console.log("result", result);
      return result;

    default:
      return state;
  }
};

const randomJokeReducer = (state: AppState = defaultState, action: Action) => {
  switch (action.type) {
    case FETCH_RANDOM_JOKE_COMPLETE:
      // add random joke to state.jokes maybe?
      console.log("state", state, "payload", action.payload);
      return state;

    default:
      return state;
  }
};

const mergeReducers = reducers => (state: AppState, action: Action) => {
  reducers.forEach(reducer => {
    state = reducer(state, action);
  });
  return state;
};

export default mergeReducers([categoriesReducer, randomJokeReducer]);
