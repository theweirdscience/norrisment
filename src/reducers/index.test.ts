import reducers from "./index";
import { Category, AppState, Joke } from "../types";
import { Action, FETCH_CATEGORIES_COMPLETE, FETCH_RANDOM_JOKE_COMPLETE, CLEAR_RANDOM_JOKE } from "../actions";

const defaultState: AppState = {
  settings: {
    requestInProgress: false
  },
  categories: []
};
const randomJoke: any = {
  icon_url: "https://assets.chucknorris.host/img/avatar/chuck-norris.png",
  id: "bKq5KbcBSq6hNVswRZYu9A",
  url: "https://api.chucknorris.io/jokes/bKq5KbcBSq6hNVswRZYu9A",
  value:
    "ghrjwgdfnbjkgdnkshx sorry Chuck Norris is informing me that the universe started 13 billion years ago fgvhlfjdunhedhgsejkwxbwgdjh"
};

const clone = (o: {}) => JSON.parse(JSON.stringify(o));

describe("Reducers", () => {
  let state = clone(defaultState);
  it("should have entry state === default state", () => {
    expect(state).toMatchObject(defaultState);
  });
  it("should reduce categories", () => {
    const result = expect(reducers(state, { type: FETCH_CATEGORIES_COMPLETE, payload: ["one", "two", "three"] }));
    result.toMatchSnapshot();
    result.not.toEqual(state);
  });
  it("should reduce joke", () => {
    const result = expect(reducers(state, { type: FETCH_RANDOM_JOKE_COMPLETE, payload: randomJoke }));
    result.toMatchSnapshot();
    result.not.toEqual(state);
  });
  it("should clear the joke", () => {
    const result = expect(reducers(state, { type: CLEAR_RANDOM_JOKE }));
    result.toMatchSnapshot();
  });
});
