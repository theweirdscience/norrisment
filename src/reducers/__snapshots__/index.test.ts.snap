// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducers should clear the joke 1`] = `
Object {
  "categories": Array [],
  "randomJoke": null,
  "settings": Object {
    "requestInProgress": false,
  },
}
`;

exports[`Reducers should reduce categories 1`] = `
Object {
  "categories": Array [
    Object {
      "icon": "",
      "name": "one",
    },
    Object {
      "icon": "",
      "name": "two",
    },
    Object {
      "icon": "",
      "name": "three",
    },
  ],
  "settings": Object {
    "requestInProgress": false,
  },
}
`;

exports[`Reducers should reduce joke 1`] = `
Object {
  "categories": Array [],
  "randomJoke": Object {
    "icon_url": "https://assets.chucknorris.host/img/avatar/chuck-norris.png",
    "id": "bKq5KbcBSq6hNVswRZYu9A",
    "url": "https://api.chucknorris.io/jokes/bKq5KbcBSq6hNVswRZYu9A",
    "value": "ghrjwgdfnbjkgdnkshx sorry Chuck Norris is informing me that the universe started 13 billion years ago fgvhlfjdunhedhgsejkwxbwgdjh",
  },
  "settings": Object {
    "requestInProgress": false,
  },
}
`;
